source/main.py

```diff
@@ -0,0 +1,307 @@
+            
+    #@trainer.on(Events.EPOCH_COMPLETED)
+    def log_training_results(engine):
+        train_evaluator.run(dtloader.train_dataloader)
+        metrics = train_evaluator.state.metrics
+        avg_accuracy = metrics['accuracy']
+        avg_bce = metrics['bce']
+        pbar.log_message(
+            "Training Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}"
+            .format(engine.state.epoch, avg_accuracy, avg_bce))
+        if log_dir:
+            writer.add_scalar("training/avg_accuracy", avg_accuracy, engine.state.epoch)
+            writer.add_scalar("training/avg_loss", avg_bce, engine.state.epoch)
+        
+    def log_validation_results(engine):
+        validation_evaluator.run(dtloader.val_dataloader)
+        metrics = validation_evaluator.state.metrics
+        avg_accuracy = metrics['accuracy']
+        avg_bce = metrics['bce']
+        avg_precision = metrics['Precision']
+        avg_recall = metrics['Recall']
+        avg_F1 = metrics['F1']
+        pbar.log_message(
+            "Validation Results - Epoch: {} Averages: Acc: {:.3f} Loss: {:.3f} Precision: {:.3f} Recall: {:.3f} F1: {:.3f}"
+            .format(engine.state.epoch, avg_accuracy, avg_bce, avg_precision, avg_recall, avg_F1))
+        pbar.n = pbar.last_print_n = 0
+        if log_dir:
+            writer.add_scalar("validation/avg_accuracy", avg_accuracy, engine.state.epoch)
+            writer.add_scalar("validation/avg_loss", avg_bce, engine.state.epoch)
+            writer.add_scalar("validation/avg_F1", avg_F1, engine.state.epoch)
+            writer.add_scalar("validation/avg_precision", avg_precision, engine.state.epoch)
+            writer.add_scalar("validation/avg_recall", avg_recall, engine.state.epoch)   
+    
+    # Events Handler (Sets when given events are happening)
+    handler = EarlyStopping(patience=3, score_function=score_function, trainer=trainer)
+    validation_evaluator.add_event_handler(Events.COMPLETED, handler)
+    
+    trainer.add_event_handler(Events.ITERATION_COMPLETED, log_training_loss)
+    trainer.add_event_handler(Events.ITERATION_COMPLETED, subset_validation_loss)
+    trainer.add_event_handler(Events.EPOCH_COMPLETED, log_training_results)
+    trainer.add_event_handler(Events.EPOCH_COMPLETED, log_validation_results)
+
+    # Checkpoints model/Adds event handler to trainer engine
+    if model_checkpoint_dir:
+        model_checkpoint_path = get_new_run_directory(model_checkpoint_dir)
+        checkpointer = ModelCheckpoint(model_checkpoint_path, 'ModelWSD', save_interval=1, n_saved=2, 
+                                    create_dir=True, save_as_state_dict=True,require_empty=False)
+        trainer.add_event_handler(Events.EPOCH_COMPLETED, checkpointer, {'ModelWSD': model})
+
+
+    # kick everything off
+    trainer.run(dtloader.train_dataloader, max_epochs=epochs)
+
+    if log_dir: writer.close()
+
+
+if __name__ == "__main__":
```

Comment:
this is basically where your main.py file should start, with function imports and other minimal imports ahead of time.